### PR TITLE
Refactor/electron.js

### PR DESCRIPTION
**Pull request by wongpratan**
Convert 
public/electron.js
public/sources/hear-this.js
public/sources/index.js
public/sources/scripture-app-builder.js
public/sources/util.js
to TypeScript

### CODE REVIEW

File: public/sources/util.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      return text;
+    });
+    // use .filter() to remove any undefined elements
+    verses = verses.filter((v: string) => v);
+    // only return 3
+    if (verses.length > 3) verses.pop();
+    return verses;
+  } catch (err) {
+    console.error('Failed to get sample verses', err);
+    return ['Failed to get sample verses'];
```
Comment (irahopkinson) [inline public/sources/util.ts:71]: Info: returning this error in a array could be mistaken for actual data. We might need to re-evaluate this down later.

File: src/App/analytics.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import { AnalyticsInterface } from '../../public/models/analytic.model';
+import isDev from 'electron-is-dev';
+
+const DEV_TRACK_ID = 'UA-169320344-1';
+const TRACK_ID = 'UA-22170471-17';
+
+export default class Analytics implements AnalyticsInterface {
+  constructor(settings: { enableAnalytics: boolean }) {
+    this.isEnabled = !isDev && settings.enableAnalytics;
+    this.ga = null;
```
Comment (irahopkinson) [inline src/App/analytics.ts:12]: It's best not to use `null` in TS - use `undefined` where possible. You can remove this assignment if you make the `ga` property optional.

File: src/App/components/FileSelector.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Button, IconName, MaybeElement } from '@blueprintjs/core';
+import './FileSelector.scss';
+import { remote, OpenDialogOptions, SaveDialogOptions } from 'electron';
+// const { remote } = window.require('electron');
```
Comment (irahopkinson) [inline src/App/components/FileSelector.tsx:6]: Remove comment.

**Review by irahopkinson — CHANGES_REQUESTED** [review]
Good work on this @wongpratan. There are a few minor things to clean. We can look at these tomorrow if you like. I have also rebased your branch ready to merge so do reset your local branch before making any changes.

File: src/App/analytics.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    updateProxy(settings.enableAnalytics);
+    reaction(
+      (): boolean => settings.enableAnalytics,
+      (enableAnalytics: boolean): void => {
+        updateProxy(enableAnalytics);
+      }
+    );
+  }
+  isEnabled: boolean;
+  ga: GoogleAnalytics | null;
```
Comment (irahopkinson) [inline src/App/analytics.ts:27]: You can make this property optional and remove the `null`.

File: public/sources/util.ts (hunk 1 above)
Comment (megahirt) [reply public/sources/util.ts:71]: right.  Probably best to have an Error type that we propagate up through the function chain.

File: public/models/analytic.model.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import GoogleAnalytics from 'electron-ga-uuid';
+
+export interface AnalyticsInterface {
+  isEnabled: boolean;
+  ga?: GoogleAnalytics | null;
```
[the remark this answers is not in the document]
Comment (megahirt) [reply public/models/analytic.model.ts]: done.

File: src/App/components/Actions.tsx [hunk 6]
```diff
@@ -55,40 +68,44 @@ const TextWrapper = styled(Flex).attrs({
...
+    const progress: Progress = appState.progress.combined === prop.combined ? appState.progress : null;
     const inProgress = _.get(progress, 'inProgress');
     return (
       <ActionButton
-        intent={_.get(progress, 'error') ? Intent.DANGER : intent}
-        disabled={disabled}
-        onClick={appState.progress.inProgress ? null : onClick}
+        intent={_.get(progress, 'error') ? Intent.DANGER : prop.intent}
+        disabled={prop.disabled}
+        onClick={appState.progress.inProgress ? (): void => { return; } : prop.onClick}
```
[the remark this answers is not in the document]
Comment (megahirt) [reply src/App/components/Actions.tsx]: done.

File: src/App/components/Analytics.tsx [hunk 7]
```diff
@@ -31,19 +46,20 @@ export const AnalyticsProvider = ({ settings, children }) => {
...
+}
 
 AnalyticsProvider.propTypes = {
   settings: PropTypes.object,
   children: PropTypes.node,
 };
 
-export const useAnalytics = () => {
+export function useAnalytics(): AnalyticsContext {
+  // analytics
```
[the remark this answers is not in the document]
Comment (megahirt) [reply src/App/components/Analytics.tsx]: done.

File: src/App/components/Analytics.tsx [hunk 8]
```diff
@@ -3,17 +3,32 @@ import PropTypes from 'prop-types';
...
+interface AnalyticsProviderSettings {
+  enableAnalytics: boolean;
+  setEnableAnalytics(confirmed: boolean): void;
+}
+
+const StyleText = styled(Text).attrs({
+  mb: 2,
+})``;
+
+export function AnalyticsProvider(settings: AnalyticsProviderSettings, children: any): JSX.Element {
```
[the remark this answers is not in the document]
Comment (megahirt) [reply src/App/components/Analytics.tsx]: done.

File: src/App/components/ColorPicker.tsx [hunk 9]
```diff
@@ -64,9 +64,7 @@ ColorPicker.defaultProps = {
   color: undefined,
   presetColors: SWATCH_COLORS,
   disableAlpha: true,
-  onChange: (color) => {
-    this.setState({ color: color.rgb });
-  },
+  onChange: (color: { rgb: string }): void => {
+    // this.setState({ color: color.rgb });
```
[the remark this answers is not in the document]
Comment (megahirt) [reply src/App/components/ColorPicker.tsx]: uncommented.  done.

File: public/electron.js [hunk 10]
```diff
@@ -1,129 +0,0 @@
...
-    const projects = flatten(
-      map(rootDirectories, (directories, projectType) => {
-        return sources[projectType].getProjectStructure(directories);
-      })
-    );
-    event.sender.send('did-finish-getprojectstructure', projects);
-  });
-}
-
-function handleSubmission() {
```
[the remark this answers is not in the document]
Comment (megahirt) [reply public/electron.js:83]: The function needs to be re-implemented, however this PR deals only with converting various files to TS.  Once we get this merged, a separate task will be to implement this method.  For now, I've left it as a stub method with a TODO.

File: src/App/components/FileSelector.tsx (hunk 3 above)
Comment (megahirt) [reply src/App/components/FileSelector.tsx]: removed.

File: src/App/analytics.ts (hunk 4 above)
Comment (megahirt) [reply src/App/analytics.ts]: I don't think you can make a class member/property optional (interfaces can have optional properties).  I simply removed the `null` and removed the assignment to null above.  My IDE is happy with it.

File: src/App/analytics.ts (hunk 2 above)
Comment (megahirt) [reply src/App/analytics.ts]: removed.

File: public/sources/index.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+
+export default class SourceIndex {
+  static getProject(projectType: string): ProjectSource | null {
+    switch (projectType) {
+      case hearThis.PROJECT_TYPE:
+        return hearThis;
+      case scriptureAppBuilder.PROJECT_TYPE:
+        return scriptureAppBuilder;
+      default:
+        return null;
```
[the remark this answers is not in the document]
Comment (megahirt) [reply public/sources/index.ts]: done.

File: src/App/analytics.ts (hunk 4 above)
Comment (megahirt) [reply src/App/analytics.ts]: it turns out while you cannot have an optional class property, you are still required to initialize it.  So we initialize it via shorthand to `undefined`

**Review by megahirt — APPROVED** [review]
Ira, Pong and I have worked together to get this PR reviewed and fixed up.  Well done @wongpratan !

**Pull request merged**